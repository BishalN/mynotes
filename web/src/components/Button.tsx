import { useRouter } from "next/dist/client/router";
import React, { ButtonHTMLAttributes, DetailedHTMLProps } from "react";
import { Spinner } from "./Spinner";

type ButtonProps = DetailedHTMLProps<
  ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & { loading?: boolean };

export const Button: React.FC<ButtonProps> = ({
  children,
  className,
  loading,
  ...props
}) => {
  return (
    <button
      type="button"
      className={`flex items-center justify-center w-full bg-gray-500
  ml-2 p-2 text-white
    rounded-md hover:bg-gray-600 focus:shadow-md
     focus:ring-2 focus:ring-gray-800 ${className} `}
      {...props}
    >
      {loading ? (
        <span className={`self-start`}>
          <Spinner />
        </span>
      ) : null}
      <span>{children}</span>
    </button>
  );
};

type OauthButtonProps = {
  provider: "github" | "facebook" | "google";
};

export const OauthButton: React.FC<OauthButtonProps> = ({
  children,
  provider,
}) => {
  const router = useRouter();
  return (
    <Button
      onClick={() => {
        router.push(
          `${process.env.NEXT_PUBLIC_API_BASE_URL}/login/${provider}`
        );
      }}
    >
      {children}
    </Button>
  );
};
