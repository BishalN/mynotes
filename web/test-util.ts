import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export { customRender as render };
