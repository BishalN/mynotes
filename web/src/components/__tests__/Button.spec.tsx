import Reacr from "react";

import { render } from "../../../test-util";
import renderer from "react-test-renderer";
import { Button, OauthButton } from "../Button";

describe("Button", () => {
  it("should render correctly", () => {
    const { getByTestId } = render(<Button />);
    const component = getByTestId("button");

    expect(component).toBeVisible();
  });

  it("renders component unchanged", () => {
    const tree = renderer.create(<Button />);
    expect(tree).toMatchSnapshot();
  });

  it("renders spinner when button loading is passed", () => {
    const { getByRole } = render(<Button loading />);

    const spinner = getByRole("loading");
    expect(spinner).toBeVisible();
  });
});
