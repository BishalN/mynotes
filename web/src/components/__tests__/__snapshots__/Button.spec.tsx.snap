// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders component unchanged 1`] = `
<button
  className="flex items-center justify-center w-full  bg-gray-500
  ml-2 p-2 text-white
    rounded-md hover:bg-gray-600 focus:shadow-md
     focus:ring-2 focus:ring-gray-800  undefined "
  data-testid="button"
  type="button"
>
  <span />
</button>
`;
